### ISSUE TITLE
Download of large layers stalls (2)

### ISSUE THREAD

**Issue by igorwwwwwwwwwwwwwwwwwwww** [docker/hub-feedback#1233]
Hello,

We (Travis CI) are seeing what looks like a resurfacing of https://github.com/docker/hub-feedback/issues/1225. I'm copying the original message here:

---

### The issue

We are currently experiencing major problems pulling large images from Docker Hub to GCE or EC2. The download stalls partway through a given large layer and never finishes.

### What I do

This seems to affect any large image, but here's an example:

```
docker pull travisci/ci-amethyst:packer-1508250117-29cd77f
```

### What I expect

For the image to download.

### What I see instead

The download stalls partway through:

```
root@instance-2:~/# docker pull travisci/ci-amethyst:packer-1505160600
packer-1505160600: Pulling from travisci/ci-amethyst
48f0413f904d: Pull complete 
[...]
608eabb3bedc: Pull complete 
91fdd994aa8f: Downloading [=======================================>           ]  346.5MB/443.7MB
0c78b2fdb098: Downloading [========>                                          ]  400.1MB/2.439GB
It then stays stuck forever.
```

[This gist](https://gist.github.com/soulshake/445df928904525546fcbeda794d4417c) includes more details.

### Steps to reproduce

Create a plain Ubuntu 14.04 VM on GCE (us-central1-f) or AWS (us-east-1).

Run `docker pull travisci/ci-amethyst:packer-1508250117-29cd77f.`

This is currently reproducible 100% of the time for us, and occurs whether the image layers are retrieved via `docker pull` or `curl` (see reproduction script in gist above).

Note: I found that using "-C -" to tell curl to resume an aborted transfer works after aborting and re-running about a dozen times. ([more info](https://gist.github.com/soulshake/445df928904525546fcbeda794d4417c#workaround))

---

Another script that reproduces the issue:

```
#!/bin/sh

set -eu

IMG_NAME=travisci/ci-amethyst
IMG_LAYER=sha256:45464f14df488f191b61d8a409bed2139c88931f060ad70e589bbd10ba4ca3ee

if ! which curl >/dev/null 2>&1; then
  echo "Please ensure that curl is installed (e.g. apt-get install curl)..."
  exit 1
fi

if ! which jq >/dev/null 2>&1; then
  echo "Please ensure that jq is installed (e.g. apt-get install jq)..."
  exit 1
fi

# Get a token
TOKEN=$(curl -sSL "https://auth.docker.io/token?service=registry.docker.io&scope=repository:${IMG_NAME}:pull" | jq -r .token)

# Fetch the layer
exec curl \
  -SL \
  -H "Authorization: Bearer $TOKEN" \
  -o /dev/null \
  "https://registry-1.docker.io/v2/${IMG_NAME}/blobs/${IMG_LAYER}"
```

**Comment by thaJeztah** [comment]
Thanks for reporting; let me raise this internally

**Comment by jpetazzo** [comment]
@igorwwwwwwwwwwwwwwwwwwww Thanks for reporting!

Would you be able to get the CloudFront request IDs as well? @nickstenning provided a few last time and I think it turned out to be super useful when escalating the issue to AWS.

(Note that while I work at Docker, I'm not in the team operating the Hub, so my powers are limited; but I'll do whatever I can to move this to a swift resolution ✨ )

**Comment by jpetazzo** [comment]
@igorwwwwwwwwwwwwwwwwwwww I'm also just thinking — since the root cause is very likely with AWS again, if you have contacts with them, I recommend that you escalate the issue with them as well. /cc @soulshake @tiffanyfj

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
Here's a couple `X-Amz-Cf-Id`s:

```
t-CN2cyR1HUDkpTLuBSJIPkoC0QPqvAACHwhIRPvFFTfTgzQwVVm7Q==
LrdHN78M0rvhCAZilOQBK3Mk2VhKDG5JlJjvzJIa4xqjC-V42wacKQ==
fr5odDR4ZR36joRunE4ivrKMCI-3S8RCW6dwP4xKQ_orTqYS6LP5yA==
wHiRwRQJHpw5bANEIqkxbhAuTdrvZkfPkQ4EQUwF_NtTgeia4u8Svw==
xIChN4AA6oAg4-RvJKoJkS9lR3-FOTQheKX0UcHpMQfL8MpHnAJmEw==
```

**Comment by nandhini915** [comment]
FYI - we're following up and have opened an internal support request for further investigation.

**Comment by binman-docker** [comment]
Cloudfront has made some changes - @igorwwwwwwwwwwwwwwwwwwww can you see if the downloads are still stalling? I'm not able to reproduce on my end because it seems to be region-specific.

**Comment by soulshake** [comment]
@binman-docker Things are looking good on our end, thank you!

**Comment by maria-v** [comment]
We've also been seeing this issue again. 9/10 times it's just extremely slow but eventually finishes downloading (i.e. doesn't stall completely), but it's still very disruptive to have to wait 20+mins for the process that normally takes 15-20 seconds.

**Comment by thaJeztah** [comment]
@maria-v are you seeing this at this moment?

**Comment by maria-v** [comment]
@thaJeztah last known to me time I've seen it was around 11:45am PST.

**Comment by thaJeztah** [comment]
Thanks @maria-v - if you see it happening again, and are able to get more information (see the discussion above), that would be welcome; if I understood correctly from the team, these issues can be difficult to track (or monitor), because they're location dependent.

**Comment by zanes2016** [comment]
I believe I'm seeing this issue as well. Over the past week we've noticed that pulling images from docker hub sometimes take ~90 minutes to download. We're trying to pull images from AWS in Oregon and N. California (us-west-2 and us-west-1).

**Comment by zanes2016** [comment]
This seems to be related to the issue described in #9127

**Comment by genezys** [comment]
I am seing this issue again while pulling from OVH in France. Is there an issue again with the CDN?

What can I do to help diagnose the problem?

**Comment by genezys** [comment]
Here is an example of an image pull (not that large) that stalls:

```
# docker pull ruby:2.3.4-slim
2.3.4-slim: Pulling from library/ruby
aa18ad1a0d33: Downloading [=====================================>             ]  39.16MB/52.6MB
237a1e0ccc93: Download complete 
f3db0faddae6: Download complete 
c79fe184b93e: Downloading [======>                                            ]   4.48MB/36.16MB
234a577e1518: Download complete 
bffe34346292: Download complete 
```

**Comment by manishtomar** [comment]
Can you try with running following script with particular ruby blob?
```bash
#!/bin/bash
# call it as <script> library/mysql blobs/sha256:2a72cbf407d67c7a7a76dd48e432091678e297140dce050ad5eccad918a9f8d6
repo=$1
url=$2
token=$(curl "https://auth.docker.io/token?service=registry.docker.io&scope=repository:$repo:pull" | jq -r .token)
curl -v https://registry-1.docker.io/v2/$repo/$url -H "Authorization: Bearer $token" -L > /dev/null
```

**Comment by genezys** [comment]
@manishtomar Thank you for your script. How can I find the specific sha256 for the image `ruby:2.3.4-slim`?

**Comment by genezys** [comment]
@manishtomar Inspired by your script, I was able to call the URL `manifests/2.3.4-slim` to retrieve the list of blobs. Here is the transcript when downloading a blob that stalls:

<details>
<pre># ./registry_download library/ruby blobs/sha256:c79fe184b93e7353396073ca71e00848517d590af5a8da6d8faacbdbadada78e
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  4185    0  4185    0     0   3196      0 --:--:--  0:00:01 --:--:--  3197
* Hostname was NOT found in DNS cache
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 52.54.155.177...
* Connected to registry-1.docker.io (52.54.155.177) port 443 (#0)
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* SSLv3, TLS handshake, Client hello (1):
} [data not shown]
* SSLv3, TLS handshake, Server hello (2):
{ [data not shown]
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* SSLv3, TLS handshake, CERT (11):
{ [data not shown]
* SSLv3, TLS handshake, Server key exchange (12):
{ [data not shown]
* SSLv3, TLS handshake, Server finished (14):
{ [data not shown]
* SSLv3, TLS handshake, Client key exchange (16):
} [data not shown]
* SSLv3, TLS change cipher, Client hello (1):
} [data not shown]
* SSLv3, TLS handshake, Finished (20):
} [data not shown]
* SSLv3, TLS change cipher, Client hello (1):
{ [data not shown]
* SSLv3, TLS handshake, Finished (20):
{ [data not shown]
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* Server certificate:
* 	 subject: CN=*.docker.io
* 	 start date: 2018-07-05 00:00:00 GMT
* 	 expire date: 2019-08-05 12:00:00 GMT
* 	 subjectAltName: registry-1.docker.io matched
* 	 issuer: C=US; O=Amazon; OU=Server CA 1B; CN=Amazon
* 	 SSL certificate verify ok.
> GET /v2/library/ruby/blobs/sha256:c79fe184b93e7353396073ca71e00848517d590af5a8da6d8faacbdbadada78e HTTP/1.1
> User-Agent: curl/7.38.0
> Host: registry-1.docker.io
> Accept: */*
> Authorization: Bearer eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsIng1YyI6WyJNSUlDK2pDQ0FwK2dBd0lCQWdJQkFEQUtCZ2dxaGtqT1BRUURBakJHTVVRd1FnWURWUVFERXpzeVYwNVpPbFZMUzFJNlJFMUVVanBTU1U5Rk9reEhOa0U2UTFWWVZEcE5SbFZNT2tZelNFVTZOVkF5VlRwTFNqTkdPa05CTmxrNlNrbEVVVEFlRncweE9EQXlNVFF5TXpBMk5EZGFGdzB4T1RBeU1UUXlNekEyTkRkYU1FWXhSREJDQmdOVkJBTVRPMVpCUTFZNk5VNWFNenBNTkZSWk9sQlFTbGc2VWsxQlZEcEdWalpQT2xZMU1sTTZRa2szV2pwU1REVk9PbGhXVDBJNlFsTmFSanBHVTFRMk1JSUJJakFOQmdrcWhraUc5dzBCQVFFRkFBT0NBUThBTUlJQkNnS0NBUUVBMGtyTmgyZWxESnVvYjVERWd5Wi9oZ3l1ZlpxNHo0OXdvNStGRnFRK3VPTGNCMDRyc3N4cnVNdm1aSzJZQ0RSRVRERU9xNW5keEVMMHNaTE51UXRMSlNRdFY1YUhlY2dQVFRkeVJHUTl2aURPWGlqNFBocE40R0N0eFV6YTNKWlNDZC9qbm1YbmtUeDViOElUWXBCZzg2TGNUdmMyRFVUV2tHNy91UThrVjVPNFFxNlZKY05TUWRId1B2Mmp4YWRZa3hBMnhaaWNvRFNFQlpjWGRneUFCRWI2YkRnUzV3QjdtYjRRVXBuM3FXRnRqdCttKzBsdDZOR3hvenNOSFJHd3EwakpqNWtZbWFnWHpEQm5NQ3l5eDFBWFpkMHBNaUlPSjhsaDhRQ09GMStsMkVuV1U1K0thaTZKYVNEOFZJc2VrRzB3YXd4T1dER3U0YzYreE1XYUx3SURBUUFCbzRHeU1JR3ZNQTRHQTFVZER3RUIvd1FFQXdJSGdEQVBCZ05WSFNVRUNEQUdCZ1JWSFNVQU1FUUdBMVVkRGdROUJEdFdRVU5XT2pWT1dqTTZURFJVV1RwUVVFcFlPbEpOUVZRNlJsWTJUenBXTlRKVE9rSkpOMW82VWt3MVRqcFlWazlDT2tKVFdrWTZSbE5VTmpCR0JnTlZIU01FUHpBOWdEc3lWMDVaT2xWTFMxSTZSRTFFVWpwU1NVOUZPa3hITmtFNlExVllWRHBOUmxWTU9rWXpTRVU2TlZBeVZUcExTak5HT2tOQk5sazZTa2xFVVRBS0JnZ3Foa2pPUFFRREFnTkpBREJHQWlFQWdZTWF3Si9uMXM0dDlva0VhRjh2aGVkeURzbERObWNyTHNRNldmWTFmRTRDSVFEbzNWazJXcndiSjNmU1dwZEVjT3hNazZ1ZEFwK2c1Nkd6TjlRSGFNeVZ1QT09Il19.eyJhY2Nlc3MiOlt7InR5cGUiOiJyZXBvc2l0b3J5IiwibmFtZSI6ImxpYnJhcnkvcnVieSIsImFjdGlvbnMiOlsicHVsbCJdfV0sImF1ZCI6InJlZ2lzdHJ5LmRvY2tlci5pbyIsImV4cCI6MTUzMzkyNjkzNywiaWF0IjoxNTMzOTI2NjM3LCJpc3MiOiJhdXRoLmRvY2tlci5pbyIsImp0aSI6IlBNcFp6emNLSTZtMkp5VU5QbEdiIiwibmJmIjoxNTMzOTI2MzM3LCJzdWIiOiIifQ.SIFeBQvEz2L_NQVcbBA916xETNiK4MnYOjYbxxL7I32DwisY9C91K_wKA0siwftpQbNwDHhi5r7wEF92YPjffuh133qAEmpFI40xcyxTGMqRIz9gN8ySzLqQZiqlKXyCIB2G0lx3FP2aRPy7HZibXo3_jSPQPihp6Az7tiGLfvKeGZiNnesyBX-zqCtbNGnUPIB8a-3fs7MCnli7piD9Hp_FFJvSDR6JXZEi7dcbJvvYXyhuAa3YBN251faYOch34vgs_lr2iqGwG_siA4HaFAtTGkXCVtweRIDNyRXb7jc4xQz9IUTKJsxL1W60p5eiBgHVijKmQljPCbNHHJakqA
> 
< HTTP/1.1 307 Temporary Redirect
< Content-Type: text/html; charset=utf-8
< Docker-Distribution-Api-Version: registry/2.0
< Location: https://production.cloudflare.docker.com/registry-v2/docker/registry/v2/blobs/sha256/c7/c79fe184b93e7353396073ca71e00848517d590af5a8da6d8faacbdbadada78e/data?verify=1533929639-gGUiKujctcWSeSEqI5RZXsZfZZQ%3D
< Date: Fri, 10 Aug 2018 18:43:59 GMT
< Content-Length: 242
< Strict-Transport-Security: max-age=31536000
< 
* Ignoring the response-body
{ [data not shown]
100   242  100   242    0     0    200      0  0:00:01  0:00:01 --:--:--   200
* Connection #0 to host registry-1.docker.io left intact
* Issue another request to this URL: 'https://production.cloudflare.docker.com/registry-v2/docker/registry/v2/blobs/sha256/c7/c79fe184b93e7353396073ca71e00848517d590af5a8da6d8faacbdbadada78e/data?verify=1533929639-gGUiKujctcWSeSEqI5RZXsZfZZQ%3D'
* Hostname was NOT found in DNS cache
*   Trying 104.18.124.25...
* Connected to production.cloudflare.docker.com (104.18.124.25) port 443 (#1)
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* SSLv3, TLS handshake, Client hello (1):
} [data not shown]
* SSLv3, TLS handshake, Server hello (2):
{ [data not shown]
* SSLv3, TLS handshake, CERT (11):
{ [data not shown]
* SSLv3, TLS handshake, Server key exchange (12):
{ [data not shown]
* SSLv3, TLS handshake, Server finished (14):
{ [data not shown]
* SSLv3, TLS handshake, Client key exchange (16):
} [data not shown]
* SSLv3, TLS change cipher, Client hello (1):
} [data not shown]
* SSLv3, TLS handshake, Finished (20):
} [data not shown]
* SSLv3, TLS change cipher, Client hello (1):
{ [data not shown]
* SSLv3, TLS handshake, Finished (20):
{ [data not shown]
* SSL connection using TLSv1.2 / ECDHE-ECDSA-AES128-GCM-SHA256
* Server certificate:
* 	 subject: OU=Domain Control Validated; OU=PositiveSSL Multi-Domain; CN=ssl870612.cloudflaressl.com
* 	 start date: 2018-03-06 00:00:00 GMT
* 	 expire date: 2019-02-22 23:59:59 GMT
* 	 subjectAltName: production.cloudflare.docker.com matched
* 	 issuer: C=GB; ST=Greater Manchester; L=Salford; O=COMODO CA Limited; CN=COMODO ECC Domain Validation Secure Server CA 2
* 	 SSL certificate verify ok.
> GET /registry-v2/docker/registry/v2/blobs/sha256/c7/c79fe184b93e7353396073ca71e00848517d590af5a8da6d8faacbdbadada78e/data?verify=1533929639-gGUiKujctcWSeSEqI5RZXsZfZZQ%3D HTTP/1.1
> User-Agent: curl/7.38.0
> Host: production.cloudflare.docker.com
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Fri, 10 Aug 2018 18:43:59 GMT
< Content-Type: application/octet-stream
< Content-Length: 36159518
< Connection: keep-alive
< Set-Cookie: __cfduid=d48656d17d28a8aa8808ac40da4ecf59a1533926639; expires=Sat, 10-Aug-19 18:43:59 GMT; path=/; domain=.production.cloudflare.docker.com; HttpOnly; Secure
< Accept-Ranges: bytes
< Cache-Control: public, max-age=14400
< CF-Cache-Status: HIT
< CF-RAY: 44849c77f85eb777-CDG
< ETag: "2cdae5ce9643eb1c3ec0fb2270419530-2"
< Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
< Expires: Fri, 10 Aug 2018 22:43:59 GMT
< Last-Modified: Thu, 14 Sep 2017 03:56:17 GMT
< Set-Cookie: __cfduid=d48656d17d28a8aa8808ac40da4ecf59a1533926639; expires=Sat, 10-Aug-19 18:43:59 GMT; path=/; domain=.production.cloudflare.docker.com; HttpOnly; Secure
< Vary: Accept-Encoding
< x-amz-id-2: 3tQTItTT8lnzDIuRvW4uK+t6dLYHNEas4AHwuoaU/A+KxQ4dB9jRmCg8o3jyZLusgDW6rMbTfMo=
< x-amz-request-id: 9A9920914C234E3B
< x-amz-version-id: 4bp0NGhOCPcvI1sYV.VAKYZUhmP9WgoN
* Server cloudflare is not blacklisted
< Server: cloudflare
< 
{ [data not shown]
  1 34.4M    1  382k    0     0   4060      0  2:28:26  0:01:36  2:26:50     0</pre>
</details>

**Comment by manishtomar** [comment]
@genezys Interesting. Can you post your IP address? Let me talk to Cloudflare to see if something is happening on your connection.

**Comment by genezys** [comment]
@manishtomar This query should go out with the following public IP address: `51.68.81.129`.

**Comment by genezys** [comment]
@manishtomar Any news from Cloudfare?

**Comment by manishtomar** [comment]
@genezys Not yet. I've pinged them again. Hoping to get an answer soon. I am assuming this still continues to happen? Is it the same with every image?

**Comment by genezys** [comment]
@manishtomar Thank you for the news. Yes this still happening. And yes it is the same for every images that is used as a base in our services: ruby, debian, ubuntu, etc.

**Comment by genezys** [comment]
@manishtomar Is there any alternative way I can pull the missing images and push them in my Docker daemon? I am stuck right now since I cannot deploy any new services and I was just building our new infrastructure on these servers.

**Comment by manishtomar** [comment]
@genezys CF logs show the download took 4 mins is too long for 35 MB image. Can you please try again and wait for 5-10 mins to see if it completes? Also can you try from another machine possibly which has another public IP address? I suspect something is wrong with your network there. 

Regarding workaround, if you are able to get it working from another machine you could [run a pull-through cache](https://docs.docker.com/registry/recipes/mirror/) from that machine and point your daemon to that mirror. One mirror which we run is for China region: registry.docker-cn.com. See if you can download from there: `docker pull registry.docker-cn.com/library/ruby`. It might be quite slow though. You can also [try mirror.gcr.io](https://cloud.google.com/container-registry/docs/using-dockerhub-mirroring). That should be fast.

**Comment by genezys** [comment]
@manishtomar I found the issue thanks to your comment and it was my fault. I tried with another public IP address on the same infrastructure and it worked fine. 

I then checked my gateway and it appears that I configured it badly. It was using too much CPU when forwarding packets. After disabling some hardware acceleration in the pfSense configuration (and rebooting), all went well.

Thank you very much for your time 🙏
